### DISCUSSION TITLE
Fatal Python error: Cannot recover from stack overflow

### DISCUSSION THREAD

**Discussion by Owenjones2591** [fastapi/fastapi#8734] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def loop():
    s = "ss"
    while s == "ss":
        error = False
        try:
            driver.find_element_by_xpath("/html/body/div[4]/div[1]/div[3]/div/div[2]/div/button").click()
        except:
            print(f'''{Fore.RED}[-] The captcha is unsolved!{Fore.RESET}''')
            sleep(8)
            driver.refresh()
            error = True
        if not error:
            s = "notss"
    try:
        sleep(2)
        driver.find_element_by_xpath('//*[@id="sid2"]/div/form/div/input').send_keys(vidUrl)
        
        sleep(1)
        driver.find_element_by_xpath('//*[@id="sid2"]/div/form/div/div/button').click()
        
        sleep(5)
        driver.find_element_by_xpath('//*[@id="c2VuZE9nb2xsb3dlcnNfdGlrdG9r"]/div[1]/div/form/button').click()
        
        sleep(6)
        hearts = driver.find_element_by_xpath('//*[@id="c2VuZE9nb2xsb3dlcnNfdGlrdG9r"]/span').text.split()
        
        int(hearts[0])
        print(f'''{Fore.GREEN}[+] Hearts sent!{Fore.RESET}''')
        
        sleep(5)
        driver.refresh()
        sleep(10)  
        loop()
    except:
        print(f'''{Fore.RED}[-] An error occured, Retrying..{Fore.RESET}''')
        driver.refresh()
        sleep(5)
        loop()
```

### Description

I believe its because i have used loop a but i am unsure how to stop it from causing this error

### Operating System

Windows

### Operating System Details

21H2

### FastAPI Version

0.71.0

### Python Version

3.9.0

**Comment by smahm006** [comment]
Noob here but not sure if this has to do with FastAPI. Looking on Google I found:

https://stackoverflow.com/questions/50508591/fatal-python-error-cannot-recover-from-stack-overflow

Maybe you are reaching the limit of the interpreter stack during that initial while loop?

**Comment by ypsah** [comment]
Agreed.

The problematic part is that your function is infinitely recursing: `loop` is unconditionally calling `loop` itself (at the end of the `try` block and at the end of the `except` block).

I recommend reading about [stack overflow](https://en.wikipedia.org/wiki/Stack_overflow) (the programming error, not the website), and [recursion in computer science](https://en.wikipedia.org/wiki/Recursion_(computer_science)).
Wikipedia is a good starting point, but if you find it a bit terse, you should also be able to find a number of articles, blog posts and educational material online.

If I understand correctly, you tried implementing an infinite loop. The idiomatic way to do this is to use a `while True` loop. Simplifying a little, this means editing your code to go from:

```python
def loop():
    do_something()
    loop()
```

To:

```python
def loop():
    while True:
        do_something()
```

**Comment by ypsah** [comment]
For the record, when raising an issue, people tend to appreciate a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example).

I understand you may be new to programming, so let me give you an example of how _I_ would have gone around to troubleshooting this issue (which is by no mean the _one true way_, merely an example):

# Step 1: put the relevant code in a separate `.py` file and run it

This would have highlighted that your example is missing some context about:
- `driver`;
- `vidUrl`;
- `Fore`;
- `sleep`.

From the way you use it, I understand that `driver` could be an instance of [`selenium`](https://www.selenium.dev)'s [`WebDriver`](https://www.selenium.dev/selenium/docs/api/py/webdriver_remote/selenium.webdriver.remote.webdriver.html?highlight=find#selenium.webdriver.remote.webdriver.WebDriver), `Fore` comes from [`colorama`](https://github.com/tartley/colorama), `sleep` is `time.sleep` from python's standard library, and `vidUrl` is probably a string containing a [URL](https://datatracker.ietf.org/doc/html/rfc3986).

```python
from time import sleep

from colorama import Fore
fom selenium.webdriver.remote.webdriver import WebDriver

driver = WebDriver()
vidUrl = "example.com"

def loop():
    s = "ss"
    while s == "ss":
        error = False
        try:
            driver.find_element_by_xpath("/html/body/div[4]/div[1]/div[3]/div/div[2]/div/button").click()
        except:
            print(f'''{Fore.RED}[-] The captcha is unsolved!{Fore.RESET}''')
            sleep(8)
            driver.refresh()
            error = True
        if not error:
            s = "notss"
    try:
        sleep(2)
        driver.find_element_by_xpath('//*[@id="sid2"]/div/form/div/input').send_keys(vidUrl)

        sleep(1)
        driver.find_element_by_xpath('//*[@id="sid2"]/div/form/div/div/button').click()

        sleep(5)
        driver.find_element_by_xpath('//*[@id="c2VuZE9nb2xsb3dlcnNfdGlrdG9r"]/div[1]/div/form/button').click()

        sleep(6)
        hearts = driver.find_element_by_xpath('//*[@id="c2VuZE9nb2xsb3dlcnNfdGlrdG9r"]/span').text.split()

        int(hearts[0])
        print(f'''{Fore.GREEN}[+] Hearts sent!{Fore.RESET}''')

        sleep(5)
        driver.refresh()
        sleep(10)
        loop()
    except:
        print(f'''{Fore.RED}[-] An error occured, Retrying..{Fore.RESET}''')
        driver.refresh()
        sleep(5)
        loop()

if __name__ == "__main__":
    loop()
```

Make sure that the snippet runs by itself and that it exhibits the issue you are investigating. On linux, it would mean running:

```shell
$ python reproducer.py
```

Where `reproducer.py` is the name of the file I put the code in.

If you are using an [IDE](https://en.wikipedia.org/wiki/Integrated_development_environment), refer to its documentation on how to run standalone python scripts.

> Note: the snippet I posted above does not work as-is. I suspect this is because one needs to setup a server for selenium. If that were important for the issue at hand, you would need to explain how to set that up so others can run your code and reproduce the issue. Fortunately, it's not relevant, so let's forget about it.

> Note#2: By now, it should already be clear that the issue is not related to `fastapi` as `fastapi` does not appear anywhere in our snippet and the issue still occurs. That makes stackoverflow.com a better place to ask a question.

# Step 2: remove as much code as you can

It is best to go step by step about it:
- remove a few lines;
- run the code;
- make sure it still fails in the same way.

In order, I think I would have removed:
- the calls to `sleep`: to make the code faster to run;
- lines that contain `driver`: to eliminate the dependency on `selenium`;
- lines that contain `Fore`: to eliminate the dependency on `colorama`.

That would have gotten me something like:

```python
def loop():
    s = "ss"
    while s == "ss":
        error = False
        try:
            pass
        except:
            error = True
        if not error:
            s = "notss"
    try:
        loop()
    except:
        loop()

if __name__ == "__main__":
    loop()
```

Then, if it hadn't been clear by now what the issue is, I would have removed either the `while` look of the `try`/`except` block.

Only the version with the `try`/`except` block would have kept showing the issue, so that is the one I would have kept. We are now down to:

```python
$ cat reproducer.py
def loop():
    try:
        loop()
    except:
        loop()

if __name__ == "__main__":
    loop()
```

Which, thinking about it, is equivalent to:

```python
def loop():
    loop()

if __name__ == "__main__":
    loop()
```

Somewhere along the way, you might have noticed that the error message changed and is now:

```console
$ python reproducer.py
Traceback (most recent call last):
  File "/tmp/tmp.KhplXQz0Am/reproducer.py", line 5, in <module>
    loop()
  File "/tmp/tmp.KhplXQz0Am/reproducer.py", line 2, in loop
    loop()
  File "/tmp/tmp.KhplXQz0Am/reproducer.py", line 2, in loop
    loop()
  File "/tmp/tmp.KhplXQz0Am/reproducer.py", line 2, in loop
    loop()
  [Previous line repeated 996 more times]
RecursionError: maximum recursion depth exceeded
```

It could mean that I removed too many lines and am now getting into a totally unrelated issue, but actually, if you read about it, you will find that the two errors are related and this new error message gives you more information about the issue's root cause than the old one.

# Step 3: Search online for the error message

Use your favorite search engine to research: "RecursionError: maximum recursion depth exceeded".

If you don't find any explanation about it that fits your use case or you don't understand the answers, _only then_ consider asking a question on stackoverflow.com (or the relevant github project, although in this case stackoverflow.com is the way I would go myself).

Read about [how to ask a good question](https://stackoverflow.com/help/how-to-ask) if you can, it'll help you get the best answer.

**Comment by hashjaco** [comment]
From what I can tell after reading your code, you are running an infinite recursive call that keeps adding to the runtime stack. I cannot seem to find a stop condition that returns from the recursive function. When is the function complete? What are we returning from it?

**Comment by iudeen** [comment]
How is this related to FastAPI?!🙁

**Comment by tiangolo** [comment, marked as answer]
It seems from your code that you are not using FastAPI, in that case this wouldn't be the place to ask.

Funnily enough, the right place would be on StackOverflow (the same name as your error 😅)
